File: test/zoo/teco/causal_conv1d/causal_conv1d.py
# BSD 3-Clause License
#
# Copyright (c) 2024, Tecorigin Co., Ltd.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
# 1. Redistributions of source code must retain the above copyright notice, this
#    list of conditions and the following disclaimer.
#
# 2. Redistributions in binary form must reproduce the above copyright notice,
#    this list of conditions and the following disclaimer in the documentation
#    and/or other materials provided with the distribution.
#
# 3. Neither the name of the copyright holder nor the names of its
#    contributors may be used to endorse or promote products derived from
#    this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
# encoding:utf-8


import os
import sys
import json
import math
import torch
import torch.nn.functional as F
import numpy as np
sys.path.append("../zoo/teco/")
sys.path.append("../")
from executor import *


def _causal_conv1d_fn_torch(
    out: torch.Tensor,
    conv_states: torch.Tensor,
    x: torch.Tensor,
    weight: torch.Tensor,
    query_start_loc: torch.Tensor,
    cache_indices: torch.Tensor,
    has_initial_state: torch.Tensor,
    pad_slot_id: int,
) -> None:

    weight_conv = weight.transpose(0, 1).contiguous()
    dim, width = weight_conv.shape
    batch = query_start_loc.shape[0] - 1
    dtype_in = x.dtype

    weight_3d = weight_conv.unsqueeze(1)

    if cache_indices is not None:
        cache_idx_list = cache_indices.tolist()
    else:
        cache_idx_list = list(range(batch))

    if has_initial_state is not None:
        has_init_list = has_initial_state.tolist()
    else:
        has_init_list = [False] * batch

    for i in range(batch):
        state_idx = int(cache_idx_list[i])

        if cache_indices is not None and state_idx == pad_slot_id:
            continue

        start = int(query_start_loc[i].item())
        end = int(query_start_loc[i + 1].item())
        seqlen_i = end - start

        if seqlen_i == 0:
            continue

        if x.shape[0] == dim:
            seq_x = x[:, start:end].unsqueeze(0)
        else:
            seq_x = x[start:end, :].T.contiguous().unsqueeze(0)

        if has_init_list[i]:
            prefix = conv_states[state_idx].T.unsqueeze(0)
        else:
            prefix = torch.zeros(
                1, dim, width - 1, dtype=seq_x.dtype, device=seq_x.device
            )


        full_x = torch.cat([prefix, seq_x], dim=-1)


        conv_out = F.conv1d(
            full_x.to(weight_conv.dtype),
            weight_3d,
            padding=0,
            groups=dim,
        )


        conv_out = F.silu(conv_out).to(dtype_in)

        if out.shape[0] == dim:
            out[:, start:end] = conv_out.squeeze(0)
        else:
            out[start:end, :] = conv_out.squeeze(0).T.contiguous()

        new_state = (
            full_x[:, :, -(width - 1):]
            .squeeze(0)
            .T
            .contiguous()
        )
        conv_states[state_idx] = new_state


def check_inputs(param_path, input_lists, reuse_lists, output_lists):
    if param_path == "":
        print("The path of prototxt file is empty.")
        return False
    if len(input_lists) != 6:
        print("The number of input data is wrong (expected 6: x, weight, query_start_loc, cache_indices, has_initial_state, convState).")
        return False
    if len(reuse_lists) != 0:
        print("The number of reuse data is wrong.")
        return False
    if len(output_lists) != 2:
        print("The number of output data is wrong (expected 2: out, convState).")
        return False
    return True


def test_causal_conv1d(param_path, input_lists, reuse_lists, output_lists, device):
    if not check_inputs(param_path, input_lists, reuse_lists, output_lists):
        return
    if device == "cuda":
        is_avail, used_device = is_device_available(device)
        if not is_avail:
            return

    params = read_prototxt(param_path)
    input_params = params["input"]
    output_params = params["output"]

    # Read input tensors
    x = to_tensor(input_lists[0], input_params[0], device=device)  # [total_seqlen, dim] from prototxt
    total_seq_len = x.shape[0]
    dim = x.shape[1]
    x = x.T.contiguous()  # transpose to (dim, total_seqlen) for torch baseline
    weight = to_tensor(input_lists[1], input_params[1], device=device)  # [width, dim]
    query_start_loc = to_tensor(input_lists[2], input_params[2], device=device)  # [batch+1]
    cache_indices = to_tensor(input_lists[3], input_params[3], device=device)  # [batch]
    has_initial_state = to_tensor(input_lists[4], input_params[4], device=device)  # [batch]
    conv_states = to_tensor(input_lists[5], input_params[5], device=device)  # [batch, convStateLen, dim]

    # Promote to float32 for precision
    x = x.to(torch.float32)
    weight = weight.to(torch.float32)
    conv_states = conv_states.to(torch.float32)

    # Get kernel params from prototxt
    kernel_param = params.get("tecokernel_param", {})
    causal_param = kernel_param.get("causal_conv1d_param", {})
    pad_slot_id = int(causal_param.get("pad_slot_id", -1))

    # Prepare output tensor (dim, total_seqlen) for torch baseline
    batch = query_start_loc.shape[0] - 1
    width = weight.shape[0]

    out = torch.zeros(dim, total_seq_len, dtype=torch.float32, device=x.device)

    # Call torch reference implementation
    _causal_conv1d_fn_torch(
        out, conv_states, x, weight,
        query_start_loc, cache_indices, has_initial_state, pad_slot_id,
    )

    # Transpose output back to (total_seqlen, dim) for saving
    out = out.T.contiguous()

    # Save outputs
    with open(output_lists[0], "wb") as f:
        save_tensor(f, out, output_params[0]["dtype"])
    with open(output_lists[1], "wb") as f:
        save_tensor(f, conv_states, output_params[1]["dtype"])


def parse_params(filename):
    with open(filename, "r") as f:
        params = json.load(f)
    return params


if __name__ == "__main__":
    params = parse_params(sys.argv[1])
    device = sys.argv[2]
    test_causal_conv1d(params["param_path"], params["input_lists"],
                       params["reuse_lists"], params["output_lists"], device)

File: test/zoo/teco/causal_conv1d/causal_conv1d.cpp
// BSD 3- Clause License Copyright (c) 2024, Tecorigin Co., Ltd. All rights
// reserved.
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
// Redistributions of source code must retain the above copyright notice,
// this list of conditions and the following disclaimer.
// Redistributions in binary form must reproduce the above copyright notice,
// this list of conditions and the following disclaimer in the documentation
// and/or other materials provided with the distribution.
// Neither the name of the copyright holder nor the names of its contributors
// may be used to endorse or promote products derived from this software
// without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION)
// HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
// STRICT LIABILITY,OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)  ARISING IN ANY
// WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
// OF SUCH DAMAGE.

#include <stdio.h>
#include <cstring>
#include <iostream>
#include <string>
#include "zoo/teco/convert.h"
#include "common/time.hpp"
#include "zoo/teco/causal_conv1d/causal_conv1d.h"
#include "interface/include/tecoops.h"

namespace optest {

void CausalConv1dExecutor::paramCheck() {
    int nin = parser_->inputs().size();
    int nout = parser_->outputs().size();

    // 6 inputs: x, weight, query_start_loc, cache_indices, has_initial_state, convState
    // 2 outputs: out, convState (updated)
    if (nin != 6 || nout != 2) {
        ALLOG(ERROR) << "causal_conv1d requires 6 inputs and 2 outputs, got "
                     << nin << " in, " << nout << " out.";
        throw std::invalid_argument(std::string(__FILE__) + ":" + std::to_string(__LINE__));
    }
}

void CausalConv1dExecutor::paramParse() {
    auto meta_x = parser_->input(0);
    auto meta_weight = parser_->input(1);

    // x shape: [total_seqlen, dim]
    totalSeqLen_ = meta_x->shape[0];
    dim_ = meta_x->shape[1];

    // weight shape: [width, dim]
    width_ = meta_weight->shape[0];

    // Parse kernel params from prototxt
    auto causal_param = parser_->getProtoNode()->tecokernel_param().causal_conv1d_param();
    padSlotId_ = causal_param.pad_slot_id();
    api_mode_ = causal_param.api_mode();
    actMode_ = causal_param.act_mode();
    convStateStride_ = causal_param.conv_state_stride();
    xSeqStride_ = causal_param.x_seq_stride();
    outSeqStride_ = causal_param.out_seq_stride();

    // Derive batch from query_start_loc shape
    auto meta_qsl = parser_->input(2);
    batch_ = meta_qsl->shape[0] - 1;
}

void CausalConv1dExecutor::paramGeneration() {
    x_ = dev_input[0];
    weight_ = dev_input[1];
    queryStartLoc_ = static_cast<const int *>(dev_input[2]);
    convStateIndices_ = static_cast<const int *>(dev_input[3]);
    hasInitialState_ = static_cast<const int8_t *>(dev_input[4]);

    // dev_output[0] = out, dev_output[1] = updated convState
    out_ = dev_output[0];
    convState_ = dev_output[1];

    // convState is read-write: copy initial state from input to output buffer
    int64_t convStateBytes =
        static_cast<int64_t>(batch_) * (width_ - 1) * dim_ * sizeof(uint16_t);
    checkScdaErrors(scdaMemcpy(convState_, dev_input[5], convStateBytes, MemcpyDeviceToDevice));
}

void CausalConv1dExecutor::compute() {
    checkTECOOPS(tecoopsCausalConv1d(
        handle_,
        batch_,
        totalSeqLen_,
        dim_,
        width_,
        convStateStride_,
        xSeqStride_,
        outSeqStride_,
        padSlotId_,
        api_mode_,
        actMode_,
        x_,
        convState_,
        weight_,
        out_,
        queryStartLoc_,
        convStateIndices_,
        hasInitialState_));
}

void CausalConv1dExecutor::cpuCompute() {
    pythonComputeCPU("cpu");
}

int64_t CausalConv1dExecutor::getTheoryOps() {
    int64_t ops = static_cast<int64_t>(totalSeqLen_) * dim_ * 20;
    return ops;
}

int64_t CausalConv1dExecutor::getTheoryIoSize() {
    constexpr int kHalfSize = 2;
    int64_t totalTokens = static_cast<int64_t>(totalSeqLen_);
    int64_t size = 0;
    // read: x
    size += static_cast<int64_t>(totalTokens) * dim_ * kHalfSize;
    // read: weight
    size += static_cast<int64_t>(width_) * dim_ * kHalfSize;
    // read/write: convState (batch * (width-1) * dim)
    size += static_cast<int64_t>(batch_) * (width_ - 1) * dim_ * kHalfSize * 2;
    // write: out
    size += static_cast<int64_t>(totalTokens) * dim_ * kHalfSize;
    // read: queryStartLoc, cacheIndices, hasInitialState (negligible)
    return size;
}

}  // namespace optest
